src/layer/AtmosphereLayer.js

```diff
@@ -0,0 +1,414 @@
+
+            program.setScale(gl);
+
+            gl.depthMask(false);
+            gl.frontFace(gl.CW);
+
+            this.setSkyPoints(dc, program.getAltitude());
+            this.setSkyTrianglesIndices();
+
+            this.applySkyVertices(dc);
+            this.applySkyIndices(dc);
+
+            gl.drawElements(gl.TRIANGLE_STRIP, this._skyTriStrip.length, gl.UNSIGNED_SHORT, 0);
+
+            gl.depthMask(true);
+            gl.frontFace(gl.CCW);
+        };
+
+        // Internal. Intentionally not documented.
+        AtmosphereLayer.prototype.drawGround = function (dc) {
+
+            var gl = dc.currentGlContext,
+                program = dc.findAndBindProgram(GroundProgram),
+                terrain = dc.terrain,
+                textureBound;
+
+            program.loadGlobeRadius(gl, dc.globe.equatorialRadius);
+
+            program.loadEyePoint(gl, dc.navigatorState.eyePoint);
+
+            program.loadLightDirection(gl, this.activeLightDirection);
+
+            program.setScale(gl);
+
+            // Use this layer's night image when the light location is different than the eye location.
+            if (this.nightImageSource && this.lightLocation) {
+                this._activeTexture = dc.gpuResourceCache.resourceForKey(this.nightImageSource);
+                if (!this._activeTexture) {
+                    this._activeTexture = dc.gpuResourceCache.retrieveTexture(gl, this.nightImageSource);
+                }
+                textureBound = this._activeTexture && this._activeTexture.bind(dc);
+            }
+
+            for (var idx = 0, len = terrain.tessellator.currentTiles.length; idx < len; idx++) {
+                var currentTile = terrain.tessellator.currentTiles.tileArray[idx];
+                // Use the vertex origin for the terrain tile.
+                var terrainOrigin = currentTile.referencePoint;
+                program.loadVertexOrigin(gl, terrainOrigin);
+
+                // Use the draw context's modelview projection matrix, transformed to the tile's local coordinates.
+                this._mvpMatrix.copy(dc.navigatorState.modelviewProjection);
+                this._mvpMatrix.multiplyByTranslation(terrainOrigin[0], terrainOrigin[1], terrainOrigin[2]);
+                program.loadModelviewProjection(gl, this._mvpMatrix);
+
+                // Use a tex coord matrix that registers the night texture correctly on each terrain.
+                if (textureBound) {
+                    this._texMatrix.setToUnitYFlip();
+                    this._texMatrix.multiplyByTileTransform(currentTile.sector, this._fullSphereSector);
+                    program.loadTexMatrix(gl, this._texMatrix);
+                }
+
+                // Draw the tile, multiplying the current fragment color by the program's secondary color.
+                program.loadFragMode(gl, program.FRAGMODE_GROUND_SECONDARY);
+                gl.blendFunc(gl.DST_COLOR, gl.ZERO);
```

Comment:
Use terrain.beginRendering instead of terrain.tessellator.beginRendering. Apply that pattern to endRendering as well.